CARDS-1786: Import HERACLES questionnaires

acrowthe opened the pull request:
Second speadsheet set.
Import script based on cardiac_rehab and included in heracles/clinical-data/util

acrowthe committed:
CARDS-1786: Import HERACLES questionnaires
Second speadsheet set.
Import script based on cardiac_rehab and included in heracles/clinical-data/util

acrowthe commented:
To reimport if new spreadsheet changes are required:
1. Download all the spreadsheet tabs as .csv files from excel
![image](https://user-images.githubusercontent.com/20056751/171718250-a15d3e90-cce4-451c-b11a-e56bfda477c7.png)

2. Move these files to the folder one level outside of your cards installation:
![image](https://user-images.githubusercontent.com/20056751/171718468-7a139099-4c53-45ad-ac2c-1384cf89ac3d.png)

3. Open a terminal in this folder and run `python3 ./cards/heracles-resources/clinical-data/util/import.py`. The output should be a series of python3 commands for converting the new json files to xml files if the importer ran successfully.
4. Copy+paste and run the output python3 commands. Avoid copying the final `\`:
![image](https://user-images.githubusercontent.com/20056751/171719426-5d13dd05-c0df-4941-a735-a234c1cd4af1.png)

5. Move the generated xml files from the current folder to the `cards/heracles-resources/clinical-data/src/main/resources/SLING-INF/content/Questionnaires` folder, overwriting the existing questionnaires

marta- commented on heracles-resources/clinical-data/src/main/resources/SLING-INF/content/Questionnaires/6MWT.xml:
```diff
@@ -315,7 +315,7 @@
 						<name>value</name>
 						<values>
-							<value>"High Touch"</value>
+							<value>"High Touch" AND if 6mwt_reason = "Study"</value>
```
Conditional group not imported properly

acrowthe committed:
CARDS-1786: Import HERACLES questionnaires
- Question spreadsheet adjustment for multiple conditionals on 1 question
- Improve multiple condition handling for nested conditional groups

acrowthe committed:
CARDS-1786: Import HERACLES questionnaires
- Fix conditional imports and xomputed imports

acrowthe commented:
Known issue: The 6mwt computed question `Triggered Assessment Number` is never calculated due to a bug in computed questions. This will be resolved by CARDS-1800

marta- commented:
Some things noticed after a quick run through the questionnaires:
- There are only 2 questionnaires with a visit number, is that intentional? @k-pace 
- **6MWT**
  - The first question is the only one with compact layout and looks inconsistent
  - All the questions that are multiselect should be single select. This is also true for many other questionnaires. **We may want to consider making `1` instead of `0`  the default for `maxAnswers`**.
  - Polar: what is now in the description should be answer options
- **Baseline Health Information**
  - Visit number: 1 should be computed if 1 is the only option
  - "Relevant diagnoses", "Family history" -> option "Other" reveals a conditional section; should we use our "Other" input instead? **The same situation appears in other forms.**
- **CPET: External File**
  - "Vmax File Upload" is displayed as an input, should be of type file. Also, the description is unnecessary, as the generated`AnswerInstructions` will give the proper directions.
- **Demographics**
  - Should we map "Ethnicity" to HANCESTRO or do the recepients of the data want it as is? @k-pace 
- **Laboratory Results** appear to have no questions at all
- **Physical assessments** 
  - Visit number: 1 should be computed if 1 is the only option

marta- commented:
Rebased on the latest `dev` so we can visualize the conditions better in the questionnaire wizard.

k-pace commented:
> Some things noticed after a quick run through the questionnaires:
> 
> * There are only 2 questionnaires with a visit number, is that intentional? @k-pace
No, there should be 4, however some of them should be conditional or computed based on cohort/study stream (CPET: External Files, 6MWT, CPET Interpretation, Laboratory Results)
> * **6MWT**
>   
>   * The first question is the only one with compact layout and looks inconsistent
I made every question compact in the spreadsheet if there were more than ~5 questions to conserve screen space. Feel free to adjust
>   * All the questions that are multiselect should be single select. **This is also true for many other questionnaires. We may want to consider making `1` instead of `0`  the default for `maxAnswers`**.
Agree
>   * "Polar": what is now in the description should be answer options
There is only one Polar question: Was Polar data captured? (boolean) and no description. If you're seeing something else, it sounds off
> * **Baseline Health Information**
>   
>   * "Visit number" should be `computed` and prepopulated with the value `1` if that is the only option
>   * "Relevant diagnoses", "Family history" -> option "Other" reveals a conditional section; should we use our "Other" input instead? **The same situation appears in other forms.**
Agree
> * **CPET: External File**
>   
>   * "Vmax File Upload" is displayed as an input, should be of type file. Also, the description is unnecessary, as the generated `AnswerInstructions` will give the proper directions.
Removed description from spreadsheet
> * **Demographics**
>   
>   * Should we map "Ethnicity" to HANCESTRO or do the recipients of the data want it as is? @k-pace
No, leave as is
> * **Laboratory Results** appears to have no questions at all
This is a bug
> * **Physical assessments**
>   
>   * "Visit number" should be `computed` and prepopulated with the value `1` if that is the only option

acrowthe committed:
CARDS-1786: Import HERACLES questionnaires
- Address code review comments
  - Make a lot of questions compact
  - Set default maxAnswers to 1
  - `Other` option improvements
  - Other small adjustments

acrowthe commented:
> - All the questions that are multiselect should be single select. This is also true for many other questionnaires. We may want to consider making 1 instead of 0 the default for maxAnswers.

`maxAnswers` now defaults to 1 if not otherwise specified

> - "Polar": what is now in the description should be answer options

Not seeing this issue

> - "Visit number" should be computed and prepopulated with the value 1 if that is the only option

Done for `Baseline Health Information` and `Physical Assessments`

> - "Relevant diagnoses", "Family history" -> option "Other" reveals a conditional section; should we use our "Other" input instead? The same situation appears in other forms.

Done for family history and other questions with only 1 other option. Relevant diagnoses has 2 different `other` options which open up different sections depending on which is selected: Do we want to combine both questions into the one other field?


> "Vmax File Upload" is displayed as an input, should be of type file. Also, the description is unnecessary, as the generated AnswerInstructions will give the proper directions.

Fixed

> - Laboratory Results appears to have no questions at all

Fixed, there was an issue with the study visits handling due to report and sub-report being identical

marta- commented:
> > * "Polar": what is now in the description should be answer options
> 
> There is only one Polar question: Was Polar data captured? (boolean) and no description. If you're seeing something else, it sounds off

My apologies, I mistakenly said Polar. I meant Proctor ID, the last question in 6MWT.
![image](https://user-images.githubusercontent.com/651980/173365181-af6a1923-68ff-48c1-ba35-13edf0e86e33.png)

marta- commented:
This is starting to look really nice. Outstanding issues that I can see:
* **6MWT**/Proctor ID should have answer options (see previous comment)
* After the latest commit, there are many more questions with `compact` layout that isn't appropriate for them, such as **Baseline Health Information**/Relevant diagnoses (there are others)
![image](https://user-images.githubusercontent.com/651980/173366107-8d3a4fd5-b664-4d95-add3-6daed14d0b53.png)
* Some questionnaires seem to have just one section with the same title as the form, can we get rid of it and put the questions directly in the questionnaire? Example: **Physical Assessments**
![image](https://user-images.githubusercontent.com/651980/173366757-baac4c31-1932-4c3a-bbe4-b8cca9d0d5c7.png)
* Speaking of **Physical Assessments`, whenever Visit number is `computed`, the `displayMode` should be `formatted`.

Additionally, I put up #1065 for [CARDS-1804](https://phenotips.atlassian.net/browse/CARDS-1804), which is relevant to many HERACLES questions.

acrowthe committed:
CARDS-1786: Import HERACLES questionnaires
General improvements
- Make some questions with many options not compact
- Remove some extraneous sections
- Adjust styling on computed visit number questions
- Fix Proctor ID question

marta- commented:
> * **Participant Status** - We're giving mixed messages on the Date withdrawn question: "Leave blank if NA"/"This question is mandatory". **The question should not be mandatory**. ([see creenshot](https://user-images.githubusercontent.com/651980/173887168-6a2fd56c-02dd-4566-b2d6-84837a618cb4.png))

Actually, this looks more like a study design issue. We should keep the mandatory status and remove the description instead, because we show this question conditionally only when **it is applicable**.

acrowthe committed:
CARDS-1786: Import HERACLES questionnaires
General improvements
- Made all conditionals formatted
- Removed extraneous sections for Baseline Health Information,
  Baseline Medications, Baseline Cardiovascular Parameters
- Clean up Participant Status description
- Improved CPET Interpretation conditions and units

acrowthe commented:
> CPET: External File - Is study_stream intentionally spelled like this?

That is one of the questions that will be hidden and automatically filled out once we figure out how to pass High Touch vs. Low Touch into the the forms, so it doesn't have a human readable name and doesn't need one. Changed name to `Study Stream` to match other forms in case it needs to be visible at some point in the future

> Baseline Medications still has exactly one section with the same label as the questionnaire title

Fixed. Baseline Cardiovascular Parameters had the same issue due to the name being defined in `sub-report` rather than `report`, also fixed

All other issues should be resolved as requested

marta- commented:
I made some more discoveries, please bear with me.
* **Baseline Health Information** - `Cardiac rehabilitation indication` should NOT be compact
* **CPET: External file** - conditional sections `externalfile_visit_highSection` and `externalfile_visit_lowSection`:
  * Both sections end up never being visible because their conditions' `operandA` is `cohort`, a non-existing question id. It should ve `study_stream`.
  * Both `Visit number` questions should be single select 
* **CPET Interpretation**:
  *  `Systolic BP (resting)` and `Diastolic BP (resting)` have `mmHg` as their description, it should be their unit of measurement
  * `Heart rate` has "resting" as description, making it inconsistent with the 2 above
  * Some Yes/No question have vertical layout, other have horizontal (compact) layout. Let's make everything in this form vertical for consistency.
* **Physical Assessments** - let's drop `minAnswers` from the computed question `Visit Number`, because "This question is mandatory" under it looks confusing. Same for any other computed `Visit number`

acrowthe committed:
CARDS-1786: Import HERACLES questionnaires
General Improvements
- Remove minAnswers from computed visit numbers
- Fix CPET External File broken conditional and incorrect multi select
- Fix CPET Interpretation descriptions and remove most compacts
- Fix Baseline Health Information incorrect compact

acrowthe committed:
CARDS-1786: Import HERACLES questionnaires
Fix unneeded min/max values in Baseline Cardiovascular Paramenters

acrowthe commented:
All comments outside of `Year of birth` should be resolved

marta- commented:
> Known issue: The 6mwt computed question `Triggered Assessment Number` is never calculated due to a bug in computed questions. This will be resolved by CARDS-1800

This works as expected now after rebasing on the latest `dev` which includes #1066 .
